File: go/LinkedList/2_add-two-numbers/2_add-two-numbers.go
package addtwonumbers

type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	dummyHead := &ListNode{}
	current := dummyHead
	carry := 0
	l1Value := 0
	l2Value := 0
	for l1 != nil || l2 != nil || carry != 0 {
		if l1 != nil {
			l1Value = l1.Val
		} else {
			l1Value = 0
		}
		if l2 != nil {
			l2Value = l2.Val
		} else {
			l2Value = 0
		}
		sum := l1Value + l2Value + carry
		carry = sum / 10
		newNode := &ListNode{
			Val:  (sum % 10),
			Next: nil,
		}
		current.Next = newNode
		current = newNode
		if l1 != nil {
			l1 = l1.Next
		}
		if l2 != nil {
			l2 = l2.Next
		}

	}
	return dummyHead.Next

}

File: ruby/2-Add-Two-Numbers.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode} l1
# @param {ListNode} l2
# @return {ListNode}
def add_two_numbers(l1, l2)
    dummy_head = ListNode.new()
    curr = dummy_head
    
    carry = 0
    while l1 || l2 || carry > 0
        l1_val = l1 ? l1.val : 0
        l2_val = l2 ? l2.val : 0
        
        sum = l1_val + l2_val + carry
        ones = sum % 10
        carry = sum / 10

        curr.next = ListNode.new(ones)
        
        curr = curr.next
        l1 = l1.next if l1
        l2 = l2.next if l2
    end

    dummy_head.next
end
